### PR TITLE
fix(dev): record the printed origin in dev-server.json and hint on loopback 401s

### PR DESCRIPTION
**Pull request by 3mdistal**
## What was broken  In local dev, a signed-in session dies with a silent 401 storm when the app is opened via the "other" loopback label.  Session cookies are host-scoped (RFC 6265, cookies ignore ports but not hosts), so `localhost:8081` and `127.0.0.1:8081` are different cookie jars to the browser. The dev tooling was split-brained about which label is real:  - Vite printed `Local: http://localhost:8081/` (the dev server binds all interfaces, `host: "::"` by default, which Vite displays as `localhost`) - `dev-server.json` hardcoded `http://127.0.0.1:8081` (`packages/core/src/vite/client.ts`, dev action bridge plugin)  Any CLI/agent flow that read `dev-server.json` and opened `127.0.0.1` after a human signed in on `localhost` (or vice versa) sent a token the server would accept but the browser never sent: every `/_agent-native/*` call 401'd silently, then the app hard-redirected to /sign-in. Sibling dev servers on different ports of the same app produced the same signature via the shared cookie name.  ## The fix (two layers, both dev-gated)  **1. One canonical dev origin.** The dev action discovery origin is now derived from the URL Vite actually prints (`server.resolvedUrls.local[0]`), not from a second derivation of the bind address. Printed URL === `dev-server.json` origin by construction, and the CLI/agent open paths consume that same value. `isLoopbackDevActionOrigin` in `scripts/runner.ts` (the `pnpm action` forwarding gate) now also trusts the `localhost` label; it keeps accepting `127.0.0.1` so discovery files written by older dev servers still forward. If Vite prints nothing, the plugin skips the discovery file with a loud warning and `pnpm action` degrades to its existing in-process path instead of guessing a label.  **2. Dev-only unauthenticated hint.** The auth guard reads the canonical origin through a lightweight discovery-file module, so this development diagnostic does not pull the full action bridge and its runtime dependencies into hosted server bundles. An unauthenticated loopback `/_agent-native/*` request while `isDevEnvironment()` now gets the same 401 with a one-line diagnostic (body + `x-agent-native-dev-auth-hint` header) naming the canonical origin versus the visited label:  ``` {"error":"Unauthorized","hint":"You are visiting http://127.0.0.1:8081, which has its own cookie jar; this dev server's canonical origin is http://localhost:8081, so sign in or re-open the app there."} ```  Production (`NODE_ENV=production`), non-dev, and non-loopback requests keep the bare 401 — the hint is gated on `isDevEnvironment() && isLoopbackRequest(event)` (socket-peer based, not spoofable via Host) and scoped to `/_agent-native/*` only (`/api/*` untouched). It carries no session tokens or user data. New helper: `packages/core/src/server/dev-origin-hint.ts`.  ## Proof  Focused tests (`vitest`, packages/core): 154 passed across `vite/client.spec.ts`, `server/dev-origin-hint.spec.ts`, `scripts/runner.dev-forward.spec.ts`, `server/dev-action-bridge.spec.ts`, including:  - origin derivation pins recorded origin = printed Local URL, including https and bracketed IPv6, and loud skip when nothing was printed - hint builder: flip message names both origins and stays one line; already-canonical visitor gets the sign-in-again line; no Host header never throws; no token/user data leak - source contract: the hint branch stays gated on dev + loopback + `/_agent-native/*` - runner forwarding: `http://localhost:8082` discovery origin now forwards (previously ran in-process)  `corepack pnpm guards`: all 73 checks passed. `tsc --noEmit` clean for packages/core. `oxfmt` applied.`n`nThe final bundle-boundary repair passed 30 focused discovery, hint, and forwarding tests plus the core typecheck. A broader local bridge run passed 276/277 assertions; its only failure was the existing POSIX mode-bit assertion on Windows. Linux CI is the authoritative proof for that permission check and for the function-size regression.  Real interface (templates/content, fresh DB, port 8081 with a sibling pre-fix server occupying 8080):  - **Invariant**: boot log prints `Local: http://localhost:8081/`; `.agent-native/dev-server.json` records `{"origin":"http://localhost:8081",...}` — agree with the all-interfaces bind. - **Flip, both directions**: signed in on `localhost:8081`, opened `127.0.0.1:8081` → `GET /_agent-native/org/me` returned 401 with the hint in body and header (browser fetch and `curl -i`); after signing in on `127.0.0.1:8081` and clearing the localhost cookie, visiting `localhost:8081` returned 401 with the one-line sign-in-again hint. The sibling pre-fix server on 8080 still returns a bare `{"error":"Unauthorized"}` with no hint — the before/after contrast. - **Canonical stability**: signed in once on `localhost:8081`, normal use stayed signed in across the whole flip sequence (`/_agent-native/org/me` 200), and `pnpm action list-local-component-files` forwarded over HTTP to the `localhost` origin and succeeded (while `db-query`, a core script that is never HTTP-mounted, correctly fell back in-process and hit the PGlite lock — same process state, different path).  Out of scope, as agreed: per-port cookie naming, cross-origin cookies, production code paths, Content template product changes.

### CODE REVIEW

**Review by steve8708 — APPROVED** [review]

File: packages/core/src/scripts/runner.ts [hunk 1]
```diff
@@ -423,9 +423,12 @@ export async function tryForwardToDevServer(
 function isLoopbackDevActionOrigin(origin: string): boolean {
   try {
     const url = new URL(origin);
+    // Discovery files record the URL Vite prints — `localhost` on the default
+    // wildcard bind; older dev servers recorded the 127.0.0.1 literal. Both
+    // are loopback labels for the same local server.
     return (
       url.protocol === "http:" &&
-      url.hostname === "127.0.0.1" &&
+      (url.hostname === "127.0.0.1" || url.hostname === "localhost") &&
```
[the remark this answers is not in the document]
Comment (3mdistal) [reply packages/core/src/scripts/runner.ts]: Fixed in 6c60a3790e. The forwarding allowlist now accepts HTTP or HTTPS only for `localhost`, `127.0.0.1`, and `[::1]`, while retaining the root-path/no-query/no-fragment restrictions. Focused tests now prove forwarding for `https://localhost` and bracketed IPv6 and rejection of non-loopback HTTPS origins.

File: packages/core/src/server/dev-origin-hint.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  if (!visitingOrigin) {
+    return "No valid session cookie reached this dev server; sign in again on the origin the dev server printed.";
+  }
+
+  if (new URL(visitingOrigin).hostname === "localhost") {
+    return `No valid session cookie reached the dev server at ${visitingOrigin}; sign in again on this origin.`;
+  }
+
+  const canonical = new URL(visitingOrigin);
+  canonical.hostname = "localhost";
```
[the remark this answers is not in the document]
Comment (3mdistal) [reply packages/core/src/server/dev-origin-hint.ts]: Fixed in 6c60a3790e. The auth guard now reads the canonical origin from the same `dev-server.json` record written from Vite’s printed URL and passes it to the hint builder. The builder validates that record as a root HTTP(S) loopback origin and otherwise emits the generic sign-in guidance; tests cover IPv6, missing, malformed, path-bearing, and non-loopback records.

File: packages/core/src/vite/client.ts [hunk 3]
```diff
@@ -3113,13 +3113,21 @@ function devActionBridgePlugin(): Plugin {
...
         const addr = server.httpServer?.address();
         if (!addr || typeof addr !== "object" || !addr.port) return;
-        const databaseKey = hashDatabaseKey(
-          getRuntimeDatabaseUrl("pglite:./data/pglite"),
-        );
+        // The recorded origin must be the URL Vite prints (`resolvedUrls`), not
+        // a second derivation of the bind address: the browser cookie jar keys
+        // on the exact host label, so the origin a CLI/agent flow opens and
+        // the printed origin have to be one value.
+        const printedOrigin = devActionBridgeOrigin(server.resolvedUrls);
```
[the remark this answers is not in the document]
Comment (3mdistal) [reply packages/core/src/vite/client.ts:3143]: Verified against Vite’s actual startup lifecycle: `server.listen()` installs a `prependListener("listening", ...)` that assigns `server.resolvedUrls` before calling `startServer()`. The plugin listener therefore observes the populated value. I also hardened the test to begin with `resolvedUrls: null` and populate it only before the plugin listening callback, matching that ordering.

**Comment by 3mdistal** [comment]
Addressed the incremental Builder review’s direct-HTTPS finding in 45392a990c. When `x-forwarded-proto` is absent, the hint now derives the request protocol from the validated canonical discovery origin, so a direct `https://localhost` request is recognized as canonical rather than mislabeled HTTP. The new regression test passes, and the focused four-file suite is now 149/149.

File: packages/core/src/scripts/runner.ts [hunk 4]
```diff
@@ -423,9 +423,14 @@ export async function tryForwardToDevServer(
 function isLoopbackDevActionOrigin(origin: string): boolean {
   try {
     const url = new URL(origin);
+    // Discovery files record the URL Vite prints — `localhost` on the default
+    // wildcard bind; older dev servers recorded the 127.0.0.1 literal. Both
+    // are loopback labels for the same local server.
     return (
-      url.protocol === "http:" &&
-      url.hostname === "127.0.0.1" &&
+      (url.protocol === "http:" || url.protocol === "https:") &&
```
[the remark this answers is not in the document]
Comment (3mdistal) [reply packages/core/src/scripts/runner.ts:472]: Fixed in bb8df1941f. HTTPS forwarding now uses an Undici dispatcher with certificate verification disabled only after the existing strict loopback-origin validation, so the dev token cannot be sent to a remote host. The unit test asserts HTTPS requests receive that dispatcher. I also verified the transport against a real ephemeral self-signed HTTPS server: default fetch rejected the certificate and the scoped dispatcher returned the expected response.
